File: Array and Hashing/productExceptSelf.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        val resultArray = Array<Int>(nums.size) {1}
        
        var product = 1
        nums.forEachIndexed{ index, value ->
            resultArray[index]*=product
            product*=value
        }
        
        product = 1
        for(i in nums.lastIndex downTo 0) {
            resultArray[i] *= product
            product *= nums[i]
        }
        return resultArray.toIntArray()
    }
}

File: Arrays/productExceptSelf.js
const productExceptSelf = (nums) => {
    let result = [];
    let left = 1;
    let right = 1;

    for (let i = nums.length - 1; i >= 0; i--) {
        result[i] = right;
        right = right * nums[i];
    }

    for (let i = 0; i < nums.length; i++) {
        result[i] *= left;
        left *= nums[i];
    }

    return result;
}

const nums = [1, 2, 3, 4];
console.log(productExceptSelf(nums))

//[24, 12, 8, 6]
// [ 1, 1, 2, 6 ]
// [ 24, 12, 4, 1 ]